doc/sphinx/language/core/records.rst

```diff
@@ -1,248 +1,343 @@
+.. _reuse_field_name:
 
-Let us now see the work done by the :cmd:`Record` command. First the command
-generates a variant type definition with just one constructor:
-:n:`Variant @ident {* @binder } : @sort := @ident__0 {* @binder }`.
+   .. example:: Reusing a field name in multiple records
 
-To build an object of type :token:`ident`, provide the constructor
-:n:`@ident__0` with the appropriate number of terms filling the fields of the record.
+      .. coqtop:: in
 
-.. example::
+         Module A. Record R := { f : nat }. End A.
+         Module B. Record S := { f : nat }. End B.
 
-   Let us define the rational :math:`1/2`:
+      .. coqtop:: all
 
-    .. coqtop:: in
+         Check {| A.f := 0 |}.
+         Check {| B.f := 0 |}.
 
-       Theorem one_two_irred : forall x y z:nat, x * y = 1 /\ x * z = 2 -> x = 1.
-       Admitted.
+.. _record_as_clause:
 
-       Definition half := mkRat true 1 2 (O_S 1) one_two_irred.
-       Check half.
+   .. example:: Using the "as" clause in a record definition
 
-Alternatively, the following syntax allows creating objects by using named fields, as
-shown in this grammar. The fields do not have to be in any particular order, nor do they have
-to be all present if the missing ones can be inferred or prompted for
-(see :ref:`programs`).
+      .. coqtop:: all
 
-.. coqtop:: all
+         Record MyRecord := { myfield : nat } as VarName.
+         About myfield. (* observe the MyRecord variable is named "VarName" *)
 
-  Definition half' :=
-    {| sign := true;
-       Rat_bottom_cond := O_S 1;
-       Rat_irred_cond := one_two_irred |}.
+         (* make "VarName" implicit without having to rename the variable,
+            which would be necessary without the "as" clause *)
+         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
+         Check myfield.
+         Check (myfield (VarName:={| myfield := 0 |})).
 
-The following settings let you control the display format for types:
+   .. exn:: Records declared with the keyword Record or Structure cannot be recursive.
 
-.. flag:: Printing Records
+      The record name :token:`ident` appears in the type of its fields, but uses
+      the :cmd:`Record` command. Use  the :cmd:`Inductive` or
+      :cmd:`CoInductive` command instead.
 
-   When this :term:`flag` is on (this is the default),
-   use the record syntax (shown above) as the default display format.
+   .. exn:: @ident already exists
 
-You can override the display format for specified types by adding entries to these tables:
+      The fieldname :n:`@ident` is already defined as a global.
 
-.. table:: Printing Record @qualid
+   .. warn:: @ident__1 cannot be defined because the projection @ident__2 was not defined
 
-   This :term:`table` specifies a set of qualids which are displayed as records.  Use the
-   :cmd:`Add` and :cmd:`Remove` commands to update the set of qualids.
+      The type of the projection :token:`ident` depends on previous projections which
```

Comment:
Add __1 to the name? (But I vaguely remember this would not work, maybe.)

Reply:
It works if we make it
```suggestion
      The type of the projection :n:`@ident__1` depends on previous projections which
```